File: backend/demo/src/main/java/com/jobtracker/demo/model/TimelineEvent.java
package com.jobtracker.demo.model;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

/**
 * An immutable record of a single status transition in an application's history.
 * Stored as an embedded array inside JobApplication — no separate collection needed.
 */
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class TimelineEvent {

    /** Human-readable title, e.g. "OA Completed" */
    private String title;

    /** Optional extra detail, e.g. "Status changed from OA Scheduled" */
    private String description;

    /** ISO-8601 timestamp string, e.g. "2026-06-10T14:30:00" */
    private String timestamp;
}

File: frontend/lib/models/timeline_event.dart
/// A single event in an application's status history.
class TimelineEvent {
  final String title;
  final String description;
  final String timestamp; // ISO-8601 e.g. "2026-06-10T14:30:00"

  const TimelineEvent({
    required this.title,
    required this.description,
    required this.timestamp,
  });

  factory TimelineEvent.fromJson(Map<String, dynamic> json) {
    return TimelineEvent(
      title: json['title'] as String? ?? '',
      description: json['description'] as String? ?? '',
      timestamp: json['timestamp'] as String? ?? '',
    );
  }

  Map<String, dynamic> toJson() => {
        'title': title,
        'description': description,
        'timestamp': timestamp,
      };

  /// Parses the timestamp into a [DateTime]. Returns null on parse failure.
  DateTime? get dateTime {
    try {
      return DateTime.parse(timestamp);
    } catch (_) {
      return null;
    }
  }
}
